src/Raven.Server/Utils/AffinityHelper.cs

```diff
@@ -0,0 +1,201 @@
+                {
+                    // remove the N least significant bits
+                    // we do that because it is typical that the first cores (0, 1, etc) are more
+                    // powerful and we want to keep them for other things, such as request processing
+                    currentAffinity &= currentAffinity - 1;
+                }
+            }
+            else
+            {
+                currentAffinity &= pooledThread.ThreadMask.Value;
+            }
+
+            try
+            {
+                SetThreadAffinityInternal(pooledThread, currentAffinity);
+            }
+            catch (PlatformNotSupportedException)
+            {
+                // some platforms don't support it
+            }
+            catch (Exception e)
+            {
+                if (_logger.IsOperationsEnabled)
+                    _logger.Operations("Failed to set thread affinity", e);
+            }
+        }
+
+        private static void SetThreadAffinityInternal(PoolOfThreads.PooledThread pooledThread, long affinity)
+        {
+            Debug.Assert(_affinityLocker.IsReadLockHeld || _affinityLocker.IsWriteLockHeld);
+
+            if (PlatformDetails.RunningOnMacOsx)
+            {
+                // Mac OSX threads API doesn't provide a way to set thread affinity
+                // we can use thread_policy_set which will make sure that two threads will run
+                // on different cpus, however we cannot choose which cpus will be used
+
+                // from thread_policy.h about using THREAD_AFFINITY_POLICY:
+                // This may be used to express affinity relationships between threads in
+                // the task. Threads with the same affinity tag will be scheduled to
+                // share an L2 cache if possible. That is, affinity tags are a hint to
+                // the scheduler for thread placement.
+                return;
+            }
+
+            if (PlatformDetails.RunningOnPosix == false)
+            {
+                // windows
+                pooledThread.CurrentProcessThread.ProcessorAffinity = new IntPtr(affinity);
+                return;
+            }
+
+            if (PlatformDetails.RunningOnLinux)
+            {
+                var ulongAffinity = (ulong)affinity;
+                var result = Syscall.sched_setaffinity((int)pooledThread.CurrentUnmanagedThreadId, new IntPtr(sizeof(ulong)), ref ulongAffinity);
+                if (result != 0)
+                    throw new InvalidOperationException(
+                        $"Failed to set affinity for thread: {pooledThread.CurrentUnmanagedThreadId}, " +
+                        $"affinity: {affinity}, result: {result}, error: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        internal static void RemoveCustomAffinityThread(PoolOfThreads.PooledThread pooledThread)
+        {
+            using (_affinityLocker.EnterReadLock())
```

Comment:
I really don't like the locking that we have here.

Reply:
What is the problem of using RWLS?

Reply:
moved it to `ResetThreadAffinity`

Reply:
There is no need to do _any_ locking. That is handled by the OS
We just need to handle retries.
Affinity changes can also come from outside.

Reply:
done